### PR TITLE
Psalm Integration

### LINKED ISSUE #3: Psalm integration
### Feature Request

|    Q        |   A
|------------ | ------
| QA    | yes

#### Summary

As decided during the [Technical-Steering-Committee Meeting on August 3rd, 2020](https://github.com/laminas/technical-steering-committee/blob/main/meetings/minutes/2020-08-03-TSC-Minutes.md#vote-on-static-analysis-tooling), Laminas wants to implement [vimeo/psalm](https://github.com/vimeo/psalm) in all packages.

Implementing psalm is quite easy.

#### Required
- [ ] Create a `psalm.xml` in the project root
- [ ] Copy and paste the contents from this [psalm.xml.dist](https://github.com/mezzio/repo-template/blob/6124090610ef1c6d6b0c80a063cafacc3ec68f38/psalm.xml.dist)
- [ ] Run `$ composer require --dev vimeo/psalm`
- [ ] Run `$ vendor/bin/psalm --set-baseline=psalm-baseline.xml`
- [ ] Add a composer script `static-analysis` with the command `psalm --shepherd --stats`
- [ ] Add a new line to `script:` in `.travis.yml`: `- if [[ $TEST_COVERAGE == 'true' ]]; then composer static-analysis ; fi`
- [ ] Remove phpstan from the project (`phpstan.neon.dist`, `.travis.yml` entry, `composer.json` `require-dev` and `scripts`)

##### Optional
- [ ] Fix as many psalm errors as possible.

### PR DESCRIPTION
**Pull request by timdev**
|    Q          |   A
|-------------- | ------
| QA            | yes

### Description

Psalm Integration, as requested in #3

### CODE REVIEW

File: src/FlashMessages.php [hunk 1]
```diff
@@ -155,15 +159,17 @@ public function prolongFlash(): void
         }
     }
 
-    public function prepareMessages(SessionInterface $session, string $sessionKey): void
+    public function prepareMessages(SessionInterface $session): void
```
Comment (froschdesign) [inline src/FlashMessages.php]: Caution: this class is not final and this is a change of the public API.

Comment (timdev) [reply src/FlashMessages.php]: Oh... drat.

**Comment by timdev** [comment]
Closing for now, will reopen once I properly fix the BC issue I created.

File: psalm.xml [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+>
+    <projectFiles>
+        <directory name="src"/>
+        <directory name="test"/>
+        <ignoreFiles>
+            <directory name="vendor"/>
+        </ignoreFiles>
+    </projectFiles>
+
+    <issueHandlers>
```
Comment (boesing) [inline psalm.xml]: these issue handlers can be deleted - I sadly did not updated all the issues I've created last year. sorry for that.

File: psalm.xml [hunk 3]
```diff
@@ -0,0 +1,33 @@
+<?xml version="1.0"?>
+<psalm
+        totallyTyped="true"
```
Comment (boesing) [inline psalm.xml]: should be replaced with `errorLevel="1"` as `totallyTyped` is deprecated since quite a while now

**Comment by boesing** [comment]
You can switch the PR to draft mode. Done that for you. 👍🏼

File: psalm-baseline.xml [hunk 4]
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (boesing) [inline psalm-baseline.xml]: If you manage to fix these two errors, we can avoid a baseline file in this project which would be awesome. 👍🏼

**Review by boesing — CHANGES_REQUESTED** [review]

File: psalm-baseline.xml (hunk 4 above)
Comment (boesing) [reply psalm-baseline.xml]: Maybe casting a non-native typehinted "string" argument to string might help to tell psalm whats going on 👍🏼

Comment (timdev) [reply psalm-baseline.xml]: The one in the test can be (and has been) fixed by removing that test/merging it into the dependent test. Does that seem reasonable?

The one in `FlashMessages` is trickier. It's not about not recognizing a string. Rather, psalm really doesn't like `mixed`, but the public API here explicitly accepts `mixed $value`. 

With the caveat that I'm pretty new to psalm, it seems like psalm considers `mixed` to be a kind of abomination to be exorcised, and not just a maximally-wide type (like `any` in typescript).  In this instance, psalm knows full well (according to `@psalm-trace` if you you add some) that we're assigning `mixed` to `mixed`. But it appears psalm seems to take a hard line that you shouldn't do that. See: https://psalm.dev/r/69768c9c2a

It's equally (if not more) possible that I'm missing something, and I share your dream of not needing a baseline, so I'm going to give others with more psalm-experience a change to speak up.

Comment (timdev) [reply psalm-baseline.xml]: Per our brief discussion in slack, I've suppressed this error, as it's unavoidable.

File: src/FlashMessages.php [hunk 5]
```diff
@@ -161,11 +165,17 @@ public function prepareMessages(SessionInterface $session, string $sessionKey):
...
-        $sessionMessages = $session->get($sessionKey);
-        $sessionMessages = ! is_array($sessionMessages) ? [] : $sessionMessages;
+        $sessionMessages = $this->getStoredMessages($sessionKey);
 
+        /** @var array<string,mixed> $currentMessages */
         $currentMessages = [];
         foreach ($sessionMessages as $key => $data) {
+            /**
+             * The public API of flash() and flashNow() explicitly allow calling
+             * code to pass `$value`s of any time, making this unavoidable.
```
Comment (boesing) [inline src/FlashMessages.php]: guess you mean `type` rather than `time`? 😄

Comment (timdev) [reply src/FlashMessages.php]: Indeed.

**Comment by timdev** [comment]
I've rebased this, and maneuvered around a new psalm error that cropped up in the meantime (see 2823f943).

File: src/FlashMessageMiddleware.php [hunk 6]
```diff
@@ -22,7 +22,10 @@ class FlashMessageMiddleware implements MiddlewareInterface
     /** @var string */
     private $attributeKey;
 
-    /** @var callable */
+    /**
+     * @var callable
+     * @psalm-var callable(SessionInterface, string):array | array{ 0: class-string, 1: string }
```
Comment (Ocramius) [inline src/FlashMessageMiddleware.php]: This union type seems borked - would be better to add a suppression on the assignment, than to mis-document this 🤔

File: src/FlashMessages.php [hunk 7]
```diff
@@ -161,11 +165,17 @@ public function prepareMessages(SessionInterface $session, string $sessionKey):
             return;
         }
 
-        $sessionMessages = $session->get($sessionKey);
-        $sessionMessages = ! is_array($sessionMessages) ? [] : $sessionMessages;
+        $sessionMessages = $this->getStoredMessages($sessionKey);
 
+        /** @var array<string,mixed> $currentMessages */
```
Comment (Ocramius) [inline src/FlashMessages.php:168]: Are messages generally `mixed`?

File: src/FlashMessages.php [hunk 8]
```diff
@@ -183,4 +193,14 @@ public function prepareMessages(SessionInterface $session, string $sessionKey):
...
     }
+
+    /**
+     * @return StoredMessages
+     */
+    private function getStoredMessages(?string $sessionKey = null): array
+    {
+        /** @var StoredMessages|null $messages */
+        $messages = $this->session->get($sessionKey ?? $this->sessionKey, []);
+        return ! is_array($messages) ? [] : $messages;
```
Comment (Ocramius) [inline src/FlashMessages.php]: Unsure if relevant, but we may return `$message ?? []` here

**Review by Ocramius — APPROVED** [review]
IMO good improvement - up to you if you want to adjust the last rough edges, or if we should 🚢  as-is :)

File: src/FlashMessageMiddleware.php (hunk 6 above)
Comment (timdev) [reply src/FlashMessageMiddleware.php]: I struggled with this. For some reason, psalm [doesn't like](https://github.com/vimeo/psalm/issues/6230) string literals in the second element of an array-style callable. I've removed the union and sacrificed a chicken via an explicit @psalm-var in the method body.

File: src/FlashMessages.php (hunk 7 above)
Comment (timdev) [reply src/FlashMessages.php:168]: While I just store strings there like a normal person, mezzio-flash doesn't impose any limitations. It's possible some madman is out there abusing that fact mightily :-)

File: src/FlashMessages.php (hunk 8 above)
Comment (timdev) [reply src/FlashMessages.php]: Nice catch.

File: src/FlashMessageMiddleware.php (hunk 6 above)
Comment (timdev) [reply src/FlashMessageMiddleware.php]: I've further improved this after @orklah helped me realize I was testing callability of the factory in a roundabout way.

**Comment by Ocramius** [comment]
This went completely under my radar. I rebased locally, and am now reviewing/merging. Sorry for the delay, @timdev!

**Review by Ocramius — APPROVED** [review]
:ship: thanks @timdev!

**Pull request merged**